src/importlinter/contracts/acyclic.py

```diff
@@ -0,0 +1,442 @@
+
+    if package_dependency == (importer, imported) or importer_package == imported_package:
+        return None
+
+    return package_dependency
+
+
+@dataclass(frozen=True)
+class CyclesFamilyKey:
+    parent: str
+    sibilings: tuple[str, ...]
+
+    def get_siblings_format(self) -> str:
+        return "(\n  " + "\n  ".join(f"{sibling}" for sibling in self.sorted_siblings) + "\n)"
+
+    @property
+    def sorted_siblings(self) -> tuple[str, ...]:
+        return tuple(sorted(self.sibilings))
+
+    def __hash__(self) -> int:
+        return hash((self.parent, self.sorted_siblings))
+
+
+@dataclass(frozen=True)
+class Cycle:
+    members: tuple[str, ...]
+    _family_key: Optional[CyclesFamilyKey] = None
+
+    @property
+    def family_key(self) -> CyclesFamilyKey:
+        if self._family_key is not None:
+            return self._family_key
+
+        parent = _longest_common_package(modules=self.members)
+        # If there is no common package, we assume that the cycle is formed between root packages
+        if parent is None:
+            parent = _PARENT_PACKAGE_FOR_MULTIPLE_ROOTS
+
+        sibilings_set: set[str] = set()
+        parent_nesting = parent.count(".")
+
+        for member in self.members:
+            if member.startswith(parent) and member != parent:
+                sibiling = ".".join(member.split(".")[: parent_nesting + 2])
+                sibilings_set.add(sibiling)
+            else:
+                sibilings_set.add(member.split(".")[0])
+
+        sibilings = tuple(sorted(sibilings_set))
+        return CyclesFamilyKey(parent=parent, sibilings=sibilings)
+
+    def get_members_format(self) -> str:
+        return "(\n -> " + "\n -> ".join(f"{member}" for member in self.members) + "\n)"
+
+
+@dataclass(frozen=True)
+class CyclesFamily:
+    key: CyclesFamilyKey
+    cycles: list[Cycle]
+
+
+class AcyclicContract(Contract):
+    """
+    A contract that checks whether the dependency graph of modules (or packages)
+    stick to an acyclic dependencies principle (ADP).
+
+    ADP is a software design principle defined by
+    Robert C. Martin that states that "the dependency graph of packages or components should have no cycles".
+    This implies that the dependencies form a directed acyclic graph (DAG).
+
+    The found cycles are grouped into cycle families. Cycle family aggregates all cycles that have the same
+    parent package and the cycle is formed between a particular set of sibilings.
+    The example of a cycles family inside django library:
+
+    >>>> Cycles family for parent module 'django'
+
+    Sibilings:
+    (
+      django.forms
+      django.template
+      django.utils
+    )
+
+    Number of cycles: 1
+
+    Cycle 1:
+
+    (
+      -> django.utils.dates
+      -> django.utils.translation
+      -> django.template
+      -> django.forms
+      -> django.utils.dates
+    )
+
+    <<<< Cycles family for parent module 'django'
+
+    Configuration options (all options are optional):
```

Comment:
I think it would be worth adding a required `packages` field so we can specify explicitly which packages we want to check for cycles. It should support subpackages, providing no packages are descendants of each other - for example we would be able to check that `django.db` has no cycles.

Reply:
If I understand you correctly, it is already done with `include_parents` parameter (see the example in the docstring).

Reply:
I think it's a bit confusing to have both `include_parents` and `exclude_parents` - better to have `packages` as a _required_ field I think.

Then `exclude_parents` could be called `ignore_packages` for consistency with `ignore_imports`.